plugins/linux-capture/pipewire.c

```diff
@@ -366,6 +375,157 @@ static void swap_texture_red_blue(gs_texture_t *texture)
+			    SPA_POD_CHOICE_RANGE_Rectangle(
+				    &SPA_RECTANGLE(320, 240), // Arbitrary
+				    &SPA_RECTANGLE(1, 1),
+				    &SPA_RECTANGLE(8192, 4320)),
+			    SPA_FORMAT_VIDEO_framerate,
+			    SPA_POD_CHOICE_RANGE_Fraction(
+				    &SPA_FRACTION(ovi->fps_num, ovi->fps_den),
+				    &SPA_FRACTION(0, 1), &SPA_FRACTION(360, 1)),
+			    0);
+	return spa_pod_builder_pop(b, &f[0]);
+}
+
+static uint32_t build_format_params(obs_pipewire_data *obs_pw,
+				    struct spa_pod_builder *pod_builder,
+				    const struct spa_pod ***param_list)
+{
+	uint32_t n_params = 0;
+
+	const struct spa_pod **params;
+	params =
+		bzalloc(2 * obs_pw->format_info.num * sizeof(struct spa_pod *));
+
+	for (size_t i = 0; i < obs_pw->format_info.num; i++) {
+		if (obs_pw->format_info.array[i].modifiers.num == 0 ||
+		    !check_pw_version(&obs_pw->server_version, 0, 3, 33)) {
+			continue;
+		}
+		params[n_params++] = build_format(
+			pod_builder, &obs_pw->video_info,
+			obs_pw->format_info.array[i].spa_format,
+			obs_pw->format_info.array[i].modifiers.array,
+			obs_pw->format_info.array[i].modifiers.num);
+	}
+	for (size_t i = 0; i < obs_pw->format_info.num; i++) {
+		params[n_params++] = build_format(
+			pod_builder, &obs_pw->video_info,
+			obs_pw->format_info.array[i].spa_format, NULL, 0);
+	}
+	*param_list = params;
+	return n_params;
+}
+
+static bool drm_format_available(uint32_t drm_format, uint32_t *drm_formats,
+				 size_t n_drm_formats)
+{
+	for (size_t j = 0; j < n_drm_formats; j++) {
+		if (drm_format == drm_formats[j]) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static void init_format_info(obs_pipewire_data *obs_pw)
+{
+	da_init(obs_pw->format_info);
+
+	uint32_t formats[] = {
+		SPA_VIDEO_FORMAT_BGRA,
+		SPA_VIDEO_FORMAT_RGBA,
+		SPA_VIDEO_FORMAT_BGRx,
+		SPA_VIDEO_FORMAT_RGBx,
+	};
+
+	size_t n_formats = sizeof(formats) / sizeof(formats[0]);
+
+	obs_enter_graphics();
+
+	enum gs_dmabuf_flags dmabuf_flags;
+	uint32_t *drm_formats = NULL;
+	size_t n_drm_formats;
+	bool capabilities_queried = gs_query_dmabuf_capabilities(
+		&dmabuf_flags, &drm_formats, &n_drm_formats);
+	for (size_t i = 0; i < n_formats; i++) {
+		struct format_info *info =
+			da_push_back_new(obs_pw->format_info);
+		da_init(info->modifiers);
+		info->spa_format = formats[i];
+		spa_pixel_format_to_drm_format(formats[i], &info->drm_format);
+		if (!capabilities_queried)
+			continue;
+		if (!drm_format_available(info->drm_format, drm_formats,
+					  n_drm_formats))
+			continue;
+		size_t n_modifiers;
+		uint64_t *modifiers = NULL;
+		if (gs_query_dmabuf_modifiers_for_format(
+			    info->drm_format, &modifiers, &n_modifiers)) {
+			da_push_back_array(info->modifiers, modifiers,
+					   n_modifiers);
+		}
+		bfree(modifiers);
+		if (dmabuf_flags &
+		    GS_DMABUF_FLAG_IMPLICIT_MODIFIERS_SUPPORTED) {
+			uint64_t modifier_implicit = DRM_FORMAT_MOD_INVALID;
+			da_push_back(info->modifiers, &modifier_implicit);
+		}
+	}
```

Comment:
This is super nitpicky, but please allow me to suggest the following newline changes:

```c
	enum gs_dmabuf_flags dmabuf_flags;
	uint32_t *drm_formats = NULL;
	size_t n_drm_formats;

	bool capabilities_queried = gs_query_dmabuf_capabilities(
		&dmabuf_flags, &drm_formats, &n_drm_formats);

	for (size_t i = 0; i < n_formats; i++) {
		struct format_info *info;
		uint32_t drm_format;

		if (!spa_pixel_format_to_drm_format(formats[i], &drm_format))
			continue;

		info = da_push_back_new(obs_pw->format_info);
		da_init(info->modifiers);
		info->spa_format = formats[i];
		info->drm_format = drm_format;

		if (!capabilities_queried)
			continue;

		size_t n_modifiers;
		uint64_t *modifiers = NULL;
		if (gs_query_dmabuf_modifiers_for_format(
			    info->drm_format, &modifiers, &n_modifiers)) {
			da_push_back_array(info->modifiers, modifiers,
					   n_modifiers);
		}
		bfree(modifiers);

		if (dmabuf_flags &
		    GS_DMABUF_FLAG_IMPLICIT_MODIFIERS_SUPPORTED) {
			uint64_t modifier_implicit = DRM_FORMAT_MOD_INVALID;
			da_push_back(info->modifiers, &modifier_implicit);
		}
	}
```